### PR TITLE
other sdbht improvements

### PR DESCRIPTION
**Pull request by ret2libc**

### CODE REVIEW

File: src/sdbht.c [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
 
+void sdbkv_fini(SdbKv *kv) {
+	free (kv->base.key);
+	free (kv->base.value);
+}
+
 SDB_API SdbHt* sdb_ht_new() {
-	return ht_new ((DupValue)strdup, (HtKvFreeFunc)sdbkv_free, (CalcSize)strlen);
+	SdbHt *ht = ht_new ((DupValue)strdup, (HtKvFreeFunc)sdbkv_fini, (CalcSize)strlen);
+	ht->elem_size = sizeof (SdbKv);
```
Comment (radare) [inline src/sdbht.c]: check for null

**Comment by ret2libc** [comment]
SdbHt with master:
```
insert time: 0.009232 seconds
find time: 0.070190 seconds, sum = 4738954741
insert time (2 time): 3.497370 seconds
find time (2 time): 24.897873 seconds, sum = 5000114468030
```
SdbHt with this PR:
```
insert time: 0.006166 seconds
find time: 0.051466 seconds, sum = 4738954741
insert time (2 time): 0.812717 seconds
find time (2 time): 6.450335 seconds, sum = 5000114468030
```
dict on master (it wasn't changed in this PR):
```
insert time: 0.001178 seconds
find time: 0.071153 seconds, sum = 4738954741
insert time (2 time): 1.449738 seconds
find time (2 time): 14.697669 seconds, sum = 5000114468030
```

As you can see this PR considerably improves the performance compared to previous SdbHt implementation and it makes it faster than dict too (apart from the insert time on small numbers, which of course is faster in dict because it doesn't have growing... but we are talking about 5ms more, with a gain of 0.6 seconds in insert time [0.8 vs 1.4] when the numbers grow and 8 seconds of difference in finding[6.4 vs 14.6])

These tests were done with these two programs: 
- for SdbHt
```C
#include "minunit.h"
#include <stdlib.h>
#include <sdb.h>
#include <time.h>

#define SEED 0xdeadbeef
#define INSERT_SZ   10000
#define FIND_SZ     1000000

#define INSERT_SZ2  10000000
#define FIND_SZ2    100000000

#define MOD_VAL    100000

int main(int argc, char **argv) {
	// on master this defaults to allocating 1024 elements
	// on my PR it starts with 3 elements
	SdbHt *ht = ht_new_size (1024, NULL, NULL, NULL);
	ht->cmp = NULL;
	ht->hashfn = NULL;
	ht->dupkey = NULL;
	ht->calcsizeK = NULL;
	int i;

	clock_t start, end;
	double cpu_time_used, time_taken;

	// few elements
	srand(SEED);
	start = clock();
	for (i = 0; i < INSERT_SZ; ++i) {
		int val = rand() % MOD_VAL;
		ht_insert (ht, (void *)val, (void *)(ut64)val);
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("insert time: %f seconds\n", time_taken);

	long long sum = 0;
	start = clock();
	for (i = 0; i < FIND_SZ; ++i) {
		int val = rand() % MOD_VAL;
		int res = (int)(ut64)ht_find (ht, (void *)val, NULL);
		sum += res;
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("find time: %f seconds, sum = %lld\n", time_taken, sum);
	ht_free (ht);

	// many elements
	ht = ht_new_size (1024, NULL, NULL, NULL);
	ht->cmp = NULL;
	ht->hashfn = NULL;
	ht->dupkey = NULL;
	ht->calcsizeK = NULL;
	srand(SEED);
	start = clock();
	for (i = 0; i < INSERT_SZ2; ++i) {
		int val = rand() % MOD_VAL;
		ht_insert (ht, (void *)val, (void *)(ut64)val);
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("insert time (2 time): %f seconds\n", time_taken);

	sum = 0;
	start = clock();
	for (i = 0; i < FIND_SZ2; ++i) {
		int val = rand() % MOD_VAL;
		int res = (int)(ut64)ht_find (ht, (void *)val, NULL);
		sum += res;
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("find time (2 time): %f seconds, sum = %lld\n", time_taken, sum);

	ht_free (ht);
	return 0;
}
```
- for dict
```C
#include "minunit.h"
#include <stdlib.h>
#include <sdb.h>
#include <time.h>

#define SEED 0xdeadbeef
#define INSERT_SZ   10000
#define FIND_SZ     1000000

#define INSERT_SZ2  10000000
#define FIND_SZ2    100000000

#define MOD_VAL    100000

int main(int argc, char **argv) {
	// on master this defaults to allocating 1024 elements
	// on my PR it starts with 3 elements
	dict *ht = dict_new (1024, NULL);
	int i;

	clock_t start, end;
	double cpu_time_used, time_taken;

	// few elements
	srand(SEED);
	start = clock();
	for (i = 0; i < INSERT_SZ; ++i) {
		int val = rand() % MOD_VAL;
		dict_set (ht, val, val, NULL);
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("insert time: %f seconds\n", time_taken);

	long long sum = 0;
	start = clock();
	for (i = 0; i < FIND_SZ; ++i) {
		int val = rand() % MOD_VAL;
		int res = (int)(ut64)dict_get (ht, val);
		sum += res;
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("find time: %f seconds, sum = %lld\n", time_taken, sum);
	dict_free (ht);

	// many elements
	ht = dict_new (1024, NULL);
	srand(SEED);
	start = clock();
	for (i = 0; i < INSERT_SZ2; ++i) {
		int val = rand() % MOD_VAL;
		dict_set (ht, val, val, NULL);
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("insert time (2 time): %f seconds\n", time_taken);

	sum = 0;
	start = clock();
	for (i = 0; i < FIND_SZ2; ++i) {
		int val = rand() % MOD_VAL;
		int res = (int)(ut64)dict_get (ht, val);
		sum += res;
	}
	end = clock();
	time_taken = ((double)end - start)/CLOCKS_PER_SEC; // in seconds 
	printf("find time (2 time): %f seconds, sum = %lld\n", time_taken, sum);

	dict_free (ht);
	return 0;
}
```

**Comment by radare** [comment]
Why 3 elements at the beginning?

yeah good performance improvement :) good work

**Comment by ret2libc** [comment]
> Why 3 elements at the beginning? 

sorry, that's old comment I took from my previous tests, but it's not true anymore... I started all hashtables/dicts with 1024 elements.

File: src/ht.c [hunk 2]
```diff
@@ -146,162 +149,198 @@ SDB_API void ht_free(SdbHt* ht) {
...
+	memcpy (kv_dst, kv, ht->elem_size);
+	check_growing (ht);
+	return true;
 }
 
-SDB_API bool ht_insert_kv(SdbHt *ht, HtKv *kv, bool update) {
-	return internal_ht_insert_kv (ht, kv, update);
+static bool insert_update(SdbHt *ht, const char *key, void *value, bool update) {
+	ut32 key_len = calcsize_key (ht, key);
+	HtKv* kv_dst = reserve_kv (ht, key, key_len, update);
```
Comment (ret2libc) [inline src/ht.c]: FIXME: this increments both bucket and ht counts, but it is not always right... If the element is being updated, and not inserted, the count should remain the same.

File: src/sdb.c [hunk 3]
```diff
@@ -764,14 +778,18 @@ SDB_API bool sdb_foreach(Sdb* s, SdbForeachCallback cb, void *user) {
...
-				continue;
-			}
-			if (!cb (user, sdbkv_key (kv), sdbkv_value (kv))) {
-				return sdb_foreach_end (s, false);
+	for (i = 0; i < s->ht->size; ++i) {
+		HtBucket *bt = &s->ht->table[i];
+		SdbKv *kv;
+		ut32 j;
+
+		BUCKET_FOREACH (s->ht, bt, j, kv) {
```
Comment (ret2libc) [inline src/sdb.c]: FIXME: make sure this work even when `cb` deletes the current element (AFAIK this happens in radare2 code)

File: src/ht.c (hunk 2 above)
Comment (ret2libc) [reply src/ht.c]: actually this was already handled...

File: src/ht.c [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* radare2 - BSD 3 Clause License - crowell, pancake 2016 */
+/* radare2 - BSD 3 Clause License - crowell, pancake, ret2libc 2018 */
```
Comment (radare) [inline src/ht.c]: use ranged date 2016-2018

File: src/sdb.c (hunk 3 above)
Comment (radare) [reply src/sdb.c]: BUCKET_FOREACH_SAFE?

File: test/unit/test_hash.c [hunk 5]
```diff
@@ -148,7 +153,6 @@ void free_kv(HtKv *kv) {
 	Person *p = kv->value;
 	free (p->name);
 	free (p);
-	free (kv);
```
Comment (radare) [inline test/unit/test_hash.c]: This free shuold go here

File: src/ht.c [hunk 6]
```diff
@@ -146,162 +149,198 @@ SDB_API void ht_free(SdbHt* ht) {
...
+	ut32 bucket = bucketfn (ht, key);
+	HtBucket *bt = &ht->table[bucket];
+	ut32 key_len = calcsize_key (ht, key);
+	HtKv *kv;
+	ut32 j;
+
+	BUCKET_FOREACH (ht, bt, j, kv) {
+		if (is_kv_equal (ht, key, key_len, kv)) {
+			freefn (ht, kv);
+			void *src = next_kv (ht, kv);
```
Comment (radare) [inline src/ht.c]: can be null?

File: src/ht.c [hunk 7]
```diff
@@ -146,162 +149,198 @@ SDB_API void ht_free(SdbHt* ht) {
...
-	HtKv* kv = ht_find_kv (ht, key, found);
-	return (kv && *found)? kv->value : NULL;
+	HtKv *res = ht_find_kv (ht, key, found);
+	return res ? res->value : NULL;
 }
 
 // Deletes a entry from the hash table from the key, if the pair exists.
 SDB_API bool ht_delete(SdbHt* ht, const char* key) {
-	return ht_delete_internal (ht, key, NULL);
+	ut32 bucket = bucketfn (ht, key);
```
Comment (radare) [inline src/ht.c]: no need to hold this in a variable, the return value is only used once and we should trust the return value

File: test/unit/test_hash.c (hunk 5 above)
Comment (ret2libc) [reply test/unit/test_hash.c]: No, given the optimization made, the free(kv) does not need to be done anymore in the free function provided by the user. Which actually makes sense: in RList the user does not provide a free function that frees the ListIter, that's usually freed by the list implementation itself.

**Comment by ret2libc** [comment]
Ok I've addressed the comments. I've also tested the PR with radare2 master and all tests pass except one related to `pf.`, which is just about sorting and I'll fix it in radare2 when we'll merge this.

**Comment by radare** [comment]
Looks good to merge

**Pull request merged**